cms/djangoapps/contentstore/outlines.py

```diff
@@ -79,6 +79,55 @@ def _error_for_not_sequence(section, not_sequence):
     )
 
 
+def _bubbled_up_groups_from_units(group_access_from_units):
+    """
+    Return {user_partition_id: [group_ids]} to bubble up from Units to Sequence.
+
+    This is to handle a special case: If *all* of the Units in a sequence have
+    the exact same group for a given user partition, bubble that value up to the
+    Sequence as a whole. For example, say that every Unit in a Sequence has a
+    group_access that looks like: { ENROLLMENT: [MASTERS] } (where both
+    constants are ints). In this case, an Audit user has nothing to see in the
+    Sequence at all, and it's not useful to give them an empty shell. So we'll
+    act as if the Sequence as a whole had that group setting. Note that there is
+    currently no way to set the group_access setting at the sequence level in
+    Studio, so course teams can only manipulate it for individual Units.
+    """
+    # If there are no Units, there's nothing to bubble up.
+    if not group_access_from_units:
+        return {}
+
+    def _normalize_group_access_dict(group_access):
+        return {
+            user_partition_id: sorted(group_ids)  # sorted for easier comparison
+            for user_partition_id, group_ids in group_access.items()
+        }
+
+    normalized_group_access_dicts = [
+        _normalize_group_access_dict(group_access) for group_access in group_access_from_units
+    ]
+    first_unit_group_access = normalized_group_access_dicts[0]
+    rest_of_seq_group_access_list = normalized_group_access_dicts[1:]
+
+    # If there's only a single Unit, bubble up its group_access.
+    if not rest_of_seq_group_access_list:
+        return first_unit_group_access
+
+    # Otherwise, go through the user partitions and groups in our first unit
+    # and compare them to all the other group_access dicts from the units in the
+    # rest of the sequence. Only keep the ones that match exactly and do not
+    # have empty groups.
+    common_group_access = {
+        user_partition_id: group_ids
+        for user_partition_id, group_ids in first_unit_group_access.items()
+        if group_ids and all(
+            group_ids == group_access.get(user_partition_id)
+            for group_access in rest_of_seq_group_access_list
+        )
+    }
```

Comment:
> This code means that if there's a single rogue unit that has differing group_access, no groups will be bubbled up - correct?

Yes, though the normal case is to have mixed groups like this, because the normal case is to have the same sequence available to everyone for a high stakes exam and have some mix of group access controlling Units by enrollment track. Bubbling up is the edge case.

Reply:
👍🏼